### PR TITLE
Reject unrecognized filter parameters in UI and API (#1736)

### PR DESCRIPTION
**Pull request by glennmatthews**
# Closes: #1736
# What's Changed

- Update `nautobot.utilities.filters.BaseFilterSet` so that it defaults to a filter form (not to be confused with a Nautobot `FilterForm`, at least not yet...) base class that reports as `is_valid() == False` if any unrecognized/unsupported `data` keys are passed in to initialize the form.
   - This behavior is controlled by a new config, `settings.STRICT_FILTERING`, which defaults to True but can be set to False if restoring the old behavior is desirable.
   - If `settings.STRICT_FILTERING` is False, a filterset with unknown filter parameters will report `is_valid() == True`, but will log a WARNING log message so as to raise visibility that a possibly invalid filter is being requested.
- ~Update all Nautobot `FilterSet` classes to ensure that their inner `Meta` classes inherit from `BaseFilterSet.Meta`~
    - ~This inheritance is not automatically present, unfortunately; however omitting it (as any existing plugins do) just means that a given FilterSet won't have this new data validation logic.~
    - ~Update developer documentation to state this new requirement~
- Update `nautobot.core.views.generic.ObjectListView` so that it checks whether its filterset instance reports as valid, and if it doesn't, report an error in the UI and render an empty queryset.
    - Also had to update `nautobot.dcim.views.RackElevationListView` because it overrides `ObjectListView.get()` with its own custom method.
- In order to not reject view query parameters that serve purposes other than data filtering (e.g., pagination, data exporting):
    - add `non_filter_params` class property to `ObjectListView`  (and customize this property for `RackElevationListView`)
    - add `get_filter_params` instance method to `ObjectListView`, which takes the `request.GET` parameters and removes any that appear in `self.non_filter_params`
    - use the return from `self.get_filter_params(request)` to instantiate the filterset, instead of passing all of `request.GET` into the instance (added in both `ObjectListView.get()` and `RackElevationListView.get()`)
- The REST API already automatically reports any filterset validation errors, but in order to not reject API query parameters that serve purposes other than data filtering (pagination, "brief", "include", "exclude"):
    - create a custom filter backend (subclass of the previously used `django_filters.rest_framework.backends.DjangoFilterBackend`) that removes the common REST API non-filtering parameters
    - add a corresponding custom DRF-Spectacular `DjangoFilterExtension` subclass to tie into the new filter backend (needed because DRF-Spectacular only extends base classes automatically, but not their subclasses)
    - add a custom filter backend specifically for `nautobot.extras.filters.ConfigContextQuerySetMixin` that extends the base custom filter backend to specifically handle `exclude` as a non-filtering parameter, since only API views that use this mixin need special handling of `exclude`
- Add tests to the generic test classes for API and UI list views to test:
    1. that the view can be filtered by `id` at all
    2. that the view reports an error when filtering based on an unknown filter with STRICT_FILTERING enabled
    3. that the view works normally but a warning message is logged when filtering on an unknown filter with STRICT_FILTERING disabled
- Just because it was bugging me, I added a `--skip-docs-build` parameter to the `invoke unittest` task so that you can skip this 10-20 second overhead on subsequent invocations of this task when it's not needed.

image

image

# TODO

- [x] Update example_plugin FilterSets to inherit inner Meta classes appropriately
- [x] Add new unit tests to check that various FilterSets now report as `is_valid() == False` when initialized with bad parameters
- [x] Add new unit tests to check that various UI and API list views now report an error and/or return an empty list when queried with bad parameters
- [x] Improve UI/API presentation of filter errors (getting rid of `__all__`)
- [x] Add nautobot_config toggle to disable strict filtering if necessary.
- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Example Plugin Updates (when adding/changing features)
- [x] Verify that no other `ObjectListView.get` functions are needing updates

### CODE REVIEW

**Comment by lampwins** [comment]
I like and agree with this, but is it not a breaking change? Consider I have a script which calls the API and has a typo in a filter field name. I get a response today and that might be perfectly fine.

**Comment by glennmatthews** [comment]
> I like and agree with this, but is it not a breaking change? Consider I have a script which calls the API and has a typo in a filter field name. I get a response today and that might be perfectly fine.

Arguably yes, but I can also argue that it's a bug fix - your script is not filtering in the way that you expect it to filter, because of the typo; with this change you'll discover your bug and be able to fix it. :-)

**Comment by glennmatthews** [comment]
Also, it's worth noting that at the Python code level, a filterset that reports `is_valid() == False` can still be used to construct a queryset, and that queryset will still be filtered based on the parameters that the filterset *does* understand. That's why I had to add an explicit `is_valid()` check in `ObjectListView`.

**Comment by glennmatthews** [comment]
Added a `settings.STRICT_FILTERING` toggle that can be overridden to `False` to disable this new behavior.

File: nautobot/extras/filters.py [hunk 1]
```diff
@@ -105,18 +106,20 @@ class CreatedUpdatedFilterSet(django_filters.FilterSet):
...
         filter_predicates={
-            "name": "icontains",
-            "target_url": "icontains",
-            "text": "icontains",
+            "label": "icontains",
+            "description": "icontains",
             "content_type__app_label": "icontains",
             "content_type__model": "icontains",
+            "template": "icontains",
+            "fallback_value": "icontains",
```
Comment (glennmatthews) [inline nautobot/extras/filters.py:114]: These were just plain wrong before, but we didn't have a ComputedField filters test case to detect that (which I've now added)

File: nautobot/utilities/filters.py [hunk 2]
```diff
@@ -454,6 +460,29 @@ class BaseFilterSet(django_filters.FilterSet):
...
+                            # Sure would be nice if we could instead do:
+                            # self.add_error(extra_key, "Unknown filter field") but extra_key isn't a field, so...
+                            self.add_error(None, f'Unknown filter field "{extra_key}"')
+                            result = False
+                        else:
+                            logger.warning('%s: Unknown filter field "%s"', self.__class__.__name__, extra_key)
+
+                return result
+
+        form = StrictFilterForm
```
Comment (jathanism) [inline nautobot/utilities/filters.py:484]: I see where this is going, however I am concerned about this adding the requirement to subclass `BaseFilterSet.Meta` on every subclass. I am wondering if there might be a better way to get the same behavior here such as overloading [`BaseFilterSet.get_form_class()`](https://github.com/carltongibson/django-filter/blob/78210722d920f803a8142e48969bc37f0f8324ed/django_filters/filterset.py#L247) or maybe subclassing the [filterset metaclass](https://github.com/carltongibson/django-filter/blob/78210722d920f803a8142e48969bc37f0f8324ed/django_filters/filterset.py#L65) (not to be confused with inner `Meta` class).

File: nautobot/utilities/filters.py [hunk 3]
```diff
@@ -427,11 +428,58 @@ class SearchFilter(MappedPredicatesFilterMixin, django_filters.CharFilter):
...
-    Note that filtersets inheriting from this class MUST also inherit their Meta from BaseFilterSet.Meta or a subclass.
+    - Changes the default `_meta.form` to BaseFilterSetFilterForm.
+    - Ensures that `id` is always an included filter field
+    """
+
+    def __init__(self, options=None):
+        super().__init__(options)
+        if self.fields is not None and "id" not in self.fields:
+            self.fields = ("id", *self.fields)
+        self.form = getattr(options, "form", BaseFilterSetFilterForm)
```
Comment (jathanism) [inline nautobot/utilities/filters.py:463]: This is *chef's kiss*.

File: nautobot/core/api/filter_backends.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        for non_filter_param in (
+            "brief",  # used to select NestedSerializer rather than Serializer
+            "include",  # used to include computed fields (excluded by default)
+            "limit",  # pagination
+            "offset",  # pagination
+        ):
+            try:
+                data.pop(non_filter_param)
+            except KeyError:
+                pass
```
Comment (jathanism) [inline nautobot/core/api/filter_backends.py:27]: Not that it's a big deal but if we're not doing anything with this exception, why catch it?
```suggestion
            data.pop(non_filter_param, None)
```

File: nautobot/core/views/generic.py [hunk 5]
```diff
@@ -144,6 +145,21 @@ class ObjectListView(ObjectPermissionRequiredMixin, View):
...
+    )
+
+    def get_filter_params(self, request):
+        """Helper function - take request.GET and discard any parameters that are not used for queryset filtering."""
+        filter_params = request.GET.copy()
+        for non_filter_param in self.non_filter_params:
+            try:
+                filter_params.pop(non_filter_param)
+            except KeyError:
+                pass
```
Comment (jathanism) [inline nautobot/core/views/generic.py:161]: Same feedback here: 
```suggestion
            filter_params.pop(non_filter_param, None)
```

**Review by jathanism — COMMENTED** [review]
Looking good so far.

File: nautobot/utilities/filters.py [hunk 6]
```diff
@@ -423,9 +428,58 @@ class SearchFilter(MappedPredicatesFilterMixin, django_filters.CharFilter):
...
 
 
-class BaseFilterSet(django_filters.FilterSet):
+class BaseFilterSetFilterForm(forms.Form):
+    """Base/default filter form class for BaseFilterSet."""
+
+    def is_valid(self):
+        """Extend forms.Form.is_valid() to treat any unknown form data entries as a potential validation error."""
+        result = super().is_valid()
+        for extra_key in set(self.data.keys()).difference(self.cleaned_data.keys()):
```
Comment (jathanism) [inline nautobot/utilities/filters.py:437]: Another minor nit since casting a dict to a set/list/tuple does so on the keys:

```suggestion
        for extra_key in set(self.data).difference(self.cleaned_data):
```

File: nautobot/core/api/filter_backends.py (hunk 4 above)
Comment (glennmatthews) [reply nautobot/core/api/filter_backends.py:27]: Because I forgot that you could use `pop()` that way. :-)

File: nautobot/utilities/filters.py (hunk 6 above)
Comment (glennmatthews) [reply nautobot/utilities/filters.py:437]: I think I'd prefer to keep this one as-is just to be more explicit about what is going on, if that's okay by you.

File: nautobot/utilities/filters.py (hunk 3 above)
Comment (glennmatthews) [reply nautobot/utilities/filters.py]: Hope you think so about the reworked approach in the latest diffs too. :-)

File: nautobot/core/graphql/generators.py [hunk 7]
```diff
@@ -68,6 +68,10 @@ def resolve_filter(self, *args, **kwargs):
         if not filterset_class:
             return getattr(self, field_name).all()
 
+        # Inverse of substitution logic from get_filtering_args_from_filterset() - transform "_type" back to "type"
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/graphql/generators.py:71]: This is the fix for #1589.

File: nautobot/core/api/filter_backends.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+        for non_filter_param in (
+            "api_version",  # used to select the Nautobot API version
+            "brief",  # used to select NestedSerializer rather than Serializer
+            "format",  # "json" or "api", used in the interactive HTML REST API views
+            "include",  # used to include computed fields (excluded by default)
+            "limit",  # pagination
+            "offset",  # pagination
+        ):
+            data.pop(non_filter_param, None)
```
Comment (bryanculver) [inline nautobot/core/api/filter_backends.py:26]: If I am understanding this correctly this is just removing the "approved" non-filter parameters.

Since the goal of this PR is to reject unrecognized filter parameters, we would still be passing down parameters that aren't filter parameters (as intended for the validation/rejection).

I can't think of a better phrasing/naming convention but something along the lines of `control_params` or `view_params` might make this more intuitively grok-able.

File: nautobot/docs/development/best-practices.md [hunk 9]
```diff
@@ -2,6 +2,11 @@
 
 While there are many different development interfaces in Nautobot that each expose unique functionality, there are a common set of a best practices that have broad applicability to users and developers alike. This includes elements of writing Jobs, Plugins, and scripts for execution through the `nbshell`.
 
+## Base Classes
+
+- All FilterSet classes should inherit from either `nautobot.utilities.filters.BaseFilterSet` (for bare-bones models) or `nautobot.extras.filters.NautobotFilterSet` (for models that support both change-logging and custom fields).
+- All Serializer classes should inherit from `ValidatedModelSerializer`.
+
```
Comment (bryanculver) [inline nautobot/docs/development/best-practices.md:9]: Can we add a note about overloading `non_filter_params` and best practices for that, seen here: https://github.com/nautobot/nautobot/pull/1794/files#diff-292c4e491fa8d2a6b999f52af45f55576703202fa9bd731438c815c0af769af5R396-R400

As well as the `get` method.

File: nautobot/extras/tests/test_filters.py [hunk 10]
```diff
@@ -518,14 +579,10 @@ def test_content_type_id_and_object_id(self):
         self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
 
 
-class JobFilterSetTestCase(TestCase):
+class JobFilterSetTestCase(FilterTestCases.FilterTestCase):
```
Comment (bryanculver) [inline nautobot/extras/tests/test_filters.py:582]: `Job` contains a slug although appears to be missing a test. Could this inherit from `FilterTestCases.NameSlugFilterTestCase`?

File: nautobot/extras/tests/test_filters.py [hunk 11]
```diff
@@ -711,7 +760,7 @@ def test_search(self):
         self.assertEqual(self.filterset(params, self.queryset).qs.values_list("pk", flat=True)[0], value)
 
 
-class RelationshipTestCase(TestCase):
+class RelationshipTestCase(FilterTestCases.FilterTestCase):
```
Comment (bryanculver) [inline nautobot/extras/tests/test_filters.py:763]: Same as https://github.com/nautobot/nautobot/pull/1794/files#r881896209

File: nautobot/utilities/testing/filters.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        def test_id(self):
+            """Verify that the filterset supports filtering by id."""
+            params = {"id": self.queryset.values_list("pk", flat=True)[:2]}
+            filterset = self.filterset(params, self.queryset)
+            self.assertTrue(filterset.is_valid())
+            self.assertEqual(filterset.qs.count(), 2)
+
+        def test_invalid_filter(self):
+            """Verify that the filterset reports as invalid when initialized with an unsupported filter parameter."""
+            params = {"ice_cream_flavor": ["chocolate"]}
```
Comment (bryanculver) [inline nautobot/utilities/testing/filters.py:22]: Not saying it's likely a filter set will contain an `ice_cream_flavor` filter, but could we make this obscenely long to make it really hard to have a conflict? Maybe even just a random string, say `xf7qmnb1ec`?

File: nautobot/extras/tests/test_filters.py (hunk 10 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_filters.py:582]: Will give it a try, good catch.

File: nautobot/utilities/testing/views.py [hunk 13]
```diff
@@ -705,13 +706,78 @@ class ListObjectsViewTestCase(ModelViewTestCase):
...
+                self.assertIn(self._get_url("view", instance=instance1), content, msg=content)
+                self.assertNotIn(self._get_url("view", instance=instance2), content, msg=content)
+            except NoReverseMatch:
+                pass
+
+        @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"], STRICT_FILTERING=True)
+        def test_list_objects_unknown_filter_strict_filtering(self):
+            """Verify that with STRICT_FILTERING, an unknown filter results in an error message and no matches."""
+            instance1, instance2 = self._get_queryset().all()[:2]
+            response = self.client.get(f"{self._get_url('list')}?ice_cream_flavor=chocolate")
```
Comment (bryanculver) [inline nautobot/utilities/testing/views.py:741]: Same as: https://github.com/nautobot/nautobot/pull/1794/files#r881910592

File: nautobot/utilities/testing/views.py [hunk 14]
```diff
@@ -705,13 +706,78 @@ class ListObjectsViewTestCase(ModelViewTestCase):
...
+                self.assertNotIn(self._get_url("view", instance=instance2), content, msg=content)
+            except NoReverseMatch:
+                pass
+
+        @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"], STRICT_FILTERING=False)
+        def test_list_objects_unknown_filter_no_strict_filtering(self):
+            """Verify that without STRICT_FILTERING, an unknown filter is ignored."""
+            instance1, instance2 = self._get_queryset().all()[:2]
+            with self.assertLogs("nautobot.utilities.filters") as cm:
+                response = self.client.get(f"{self._get_url('list')}?ice_cream_flavor=chocolate")
```
Comment (bryanculver) [inline nautobot/utilities/testing/views.py:760]: Same as: https://github.com/nautobot/nautobot/pull/1794/files#r881910592

File: nautobot/extras/views.py [hunk 15]
```diff
@@ -681,7 +681,7 @@ class ExportTemplateBulkDeleteView(generic.BulkDeleteView):
 
 class GitRepositoryListView(generic.ObjectListView):
     queryset = GitRepository.objects.all()
-    # filterset = filters.GitRepositoryFilterSet
+    filterset = filters.GitRepositoryFilterSet
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/extras/views.py:684]: It's been that way since GitRepositories were added. IIRC it and the line below were commented out because we don't yet have a GitRepositoryFilterForm implementation, so at the time the filterset was thought to be unneeded. I think leaving the below line in place as a reminder of absent functionality is desirable.

File: nautobot/utilities/filters.py [hunk 16]
```diff
@@ -580,6 +587,14 @@ def add_filter(cls, new_filter_name, new_filter_field):
             cls._generate_lookup_expression_filters(filter_name=new_filter_name, filter_field=new_filter_field)
         )
 
+    @classmethod
+    def get_fields(cls):
+        fields = super().get_fields()
+        if "id" not in fields and (cls._meta.exclude is None or "id" not in cls._meta.exclude):
+            # Add "id" as the first key in the `fields` OrderedDict
+            fields = OrderedDict(id=[django_filters.conf.settings.DEFAULT_LOOKUP_EXPR], **fields)
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/utilities/filters.py:595]: > Why does `id` need to be first?

In most of our existing filtersets it is, so I wanted to be consistent. There's no technical reason it has to be.

> What if it's already in `fields` but not first?

Then we leave it where it is, since presumably there's a reason for that.

> Also #1746: Why an OrderedDict?

Because the base FilterSet.get_fields() in django-filters returns an OrderedDict and I didn't want to change the API return type just in case there's something in django-filters that relies on that.

File: nautobot/utilities/testing/filters.py (hunk 12 above)
Comment (glennmatthews) [reply nautobot/utilities/testing/filters.py:22]: IMHO this is already sufficiently long and safe.

**Review by glennmatthews — COMMENTED** [review]
Replies inline.

**Review by bryanculver — APPROVED** [review]
LGTM. Don't love the naming of `non_filter_params` (see: https://github.com/nautobot/nautobot/pull/1794#discussion_r881742827) but won't hold this up.

File: nautobot/dcim/views.py [hunk 17]
```diff
@@ -392,10 +393,24 @@ class RackElevationListView(generic.ObjectListView):
     """
 
     queryset = Rack.objects.prefetch_related("role")
+    non_filter_params = (
+        *generic.ObjectListView.non_filter_params,
+        "face",  # render front or rear of racks?
+        "reverse",  # control of ordering
+    )
 
     def get(self, request):
```
Comment (bryanculver) [inline nautobot/dcim/views.py:402]: Let's make sure this is the only place we are overloading a ListView's `get()`

Comment (glennmatthews) [reply nautobot/dcim/views.py:402]: The only other place I see in the code is in `ObjectChangeListView`, and it calls `super().get(...)` so it's fine.
